cmd/mount_zfs/mount_zfs.c

```diff
@@ -47,63 +47,43 @@ libzfs_handle_t *g_zfs;
  * is expected to be of the form pool/dataset, however may also refer to
  * a block device if that device contains a valid zfs label.
  */
-static char *
-parse_dataset(char *dataset)
+static void
+parse_dataset(char (*dataset)[PATH_MAX])
 {
-	char cwd[PATH_MAX];
-	struct stat64 statbuf;
-	int error;
-	int len;
-
 	/*
 	 * We expect a pool/dataset to be provided, however if we're
 	 * given a device which is a member of a zpool we attempt to
 	 * extract the pool name stored in the label.  Given the pool
 	 * name we can mount the root dataset.
 	 */
-	error = stat64(dataset, &statbuf);
-	if (error == 0) {
+	int fd = open((const char *)dataset, O_RDONLY);
+	if (fd > 0) {
 		nvlist_t *config;
-		char *name;
-		int fd;
-
-		fd = open(dataset, O_RDONLY);
-		if (fd < 0)
-			goto out;
-
-		error = zpool_read_label(fd, &config, NULL);
+		char *name = NULL;
+		zpool_read_label(fd, &config, NULL);
```

Comment:
You're going to need to check the return value here and return on error, otherwise `config` could be undefined.

Reply:
`zpool_read_label` fails only after assigning a good state to `config`, but it looks like nobody else relies on that behavior. We can avoid adding a redundant `close(fd)` and several unnecessary exit points if we can expect a NULL `nvlist_t*` to be gracefully ignored (a quick grep shows that to be a common assumption).

Reply:
That's true, but it's also an internal implementation detail of `zpool_read_label()`.  For safety callers should assume that on error `config` may be undefined.  Not checking an error code is also the kind of thing static analyzers like to complain about.